runtime/common/src/migrations/asset_registry_xcmv3.rs

```diff
@@ -0,0 +1,232 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::Balance;
+use cfg_types::tokens::{CurrencyId, CustomMetadata};
+#[cfg(feature = "try-runtime")]
+use codec::{Decode, Encode};
+use frame_support::{
+	traits::OnRuntimeUpgrade,
+	weights::{constants::RocksDbWeight, Weight},
+	StoragePrefixedMap,
+};
+use orml_traits::asset_registry::AssetMetadata;
+use sp_std::vec::Vec;
+
+pub struct Migration<
+	T,
+	Assets,
+	const EXPECTED_MAINNET_LOC_COUNT: u32,
+	const EXPECTED_MAINNET_META_COUNT: u32,
+	const EXPECTED_TESTNET_LOC_COUNT: u32,
+	const EXPECTED_TESTNET_META_COUNT: u32,
+>(sp_std::marker::PhantomData<(T, Assets)>);
+
+impl<
+		T,
+		Assets,
+		const EXPECTED_MAINNET_LOC_COUNT: u32,
+		const EXPECTED_MAINNET_META_COUNT: u32,
+		const EXPECTED_TESTNET_LOC_COUNT: u32,
+		const EXPECTED_TESTNET_META_COUNT: u32,
+	> OnRuntimeUpgrade
+	for Migration<
+		T,
+		Assets,
+		EXPECTED_MAINNET_LOC_COUNT,
+		EXPECTED_MAINNET_META_COUNT,
+		EXPECTED_TESTNET_LOC_COUNT,
+		EXPECTED_TESTNET_META_COUNT,
+	> where
+	T: orml_asset_registry::Config,
+	<T as orml_asset_registry::Config>::Balance: From<u128>,
+	<T as orml_asset_registry::Config>::CustomMetadata: From<cfg_types::tokens::CustomMetadata>,
+	<T as orml_asset_registry::Config>::AssetId: From<cfg_types::tokens::CurrencyId>,
+	AssetMetadata<
+		<T as orml_asset_registry::Config>::Balance,
+		<T as orml_asset_registry::Config>::CustomMetadata,
+	>: From<AssetMetadata<u128, cfg_types::tokens::CustomMetadata>>,
+	Assets: AssetsToMigrate,
+{
+	fn on_runtime_upgrade() -> Weight {
+		log::info!("AssetRegistryMultilocationToXCMV3: on_runtime_upgrade: started");
+		// Complexity: 2 reads
+		let (loc_count, meta_count) = Self::get_key_counts();
+
+		// Complexity: O(loc_count) writes
+		let result =
+			orml_asset_registry::LocationToAssetId::<T>::clear(loc_count.saturating_add(2), None);
+		match result.maybe_cursor {
+			None => log::info!("Cleared all LocationToAssetId entries successfully"),
+			Some(_) => {
+				log::error!(
+					"LocationToAssetId not fully cleared: {:?} remaining",
+					orml_asset_registry::LocationToAssetId::<T>::iter_keys().count()
+				)
+			}
+		}
+		log::info!(
+            "AssetRegistryMultilocationToXCMV3: LocationToAssetId clearing iteration result. backend: {} unique: {} loops: {}",
+            result.backend,
+            result.unique,
+            result.loops,
+        );
+
+		// Complexity: O(meta_count) writes
+		let result = orml_asset_registry::Metadata::<T>::clear(meta_count.saturating_add(1), None);
```

Comment:
Why the plus 1?

Reply:
See [above](https://github.com/centrifuge/centrifuge-chain/pull/1528/files#r1315426451).